### PR TITLE
[JFR] add support for opto object allocations sampling

### PR DESCRIPTION
**Pull request by D-D-H**
Summary: add support for opto object allocations sampling

Test Plan: test/jfr/event/objectsprofiling/TestOptoObjectAllocationsSampling.java

Reviewers: zhengxiaolinX, luchsh, sanhong

Issue: https://github.com/alibaba/dragonwell8/issues/64

### CODE REVIEW

File: src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp [hunk 1]
```diff
@@ -106,21 +106,32 @@ int write__artifact__klass(JfrCheckpointWriter* writer, JfrArtifactSet* artifact
...
   const traceid symbol_id = artifacts->mark(klass);
   assert(symbol_id > 0, "need to have an address for symbol!");
   writer->write(TRACE_ID(klass));
   writer->write(cld_id(klass->class_loader_data()));
   writer->write((traceid)CREATE_SYMBOL_ID(symbol_id));
   writer->write((s4)klass->access_flags().get_flags());
+  if (klass->oop_is_array()) {
+    // The object array size can not be determined statically from klass.
+    // It is determined by the elements length in object layout.
+    // So we put a magic code here to make ignore it.
```
Comment (luchsh) [inline src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp]: to make `<whom?>` to ignore it ?

File: src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp [hunk 2]
```diff
@@ -106,21 +106,32 @@ int write__artifact__klass(JfrCheckpointWriter* writer, JfrArtifactSet* artifact
...
+    writer->write((s4)ARRAY_OBJECT_SIZE_MAGIC_CODE);
```
Comment (luchsh) [inline src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp]: `MAGIC_CODE` ---> `PLACE_HOLDER`?

File: src/share/vm/jfr/objectprofiler/objectProfiler.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    if (!_sample_array_obj_alloc) {
+      _sample_array_obj_alloc = true;
+    }
+  } else {
+    ShouldNotReachHere();
+  }
+  if (_enabled == JNI_TRUE) {
+    return;
+  }
+  OrderAccess::release_store((volatile jint*)&_enabled, JNI_TRUE);
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply src/share/vm/jfr/objectprofiler/objectProfiler.cpp]: this method is called by jdk under synchronized

File: src/share/vm/jfr/recorder/access/jfrOptionSet.cpp [hunk 4]
```diff
@@ -314,6 +349,8 @@ jboolean JfrOptionSet::_sample_protection = JNI_FALSE;
 #else
 jboolean JfrOptionSet::_sample_protection = JNI_TRUE;
 #endif
+volatile jboolean JfrOptionSet::_sample_object_allocations = JNI_FALSE;
+volatile jlong JfrOptionSet::_object_allocations_sampling_interval = 1024;
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply src/share/vm/jfr/recorder/access/jfrOptionSet.cpp]: the unit is count rather ms.

File: src/share/vm/jfr/recorder/access/jfrOptionSet.cpp [hunk 5]
```diff
@@ -368,6 +406,12 @@ bool JfrOptionSet::configure(TRAPS) {
   configure._sample_threads.set_is_set(_dcmd_sample_threads.is_set());
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply src/share/vm/jfr/recorder/access/jfrOptionSet.cpp]: like "_repository_path"...

File: src/share/vm/jfr/recorder/access/jfrThreadData.hpp [hunk 6]
```diff
@@ -51,6 +51,27 @@ class JfrThreadData {
...
+  // returned by vframeStream is always invalid. This is largely due to the oopmap that
+  // is not correctly granted ( refer to PhaseMacroExpand::expand_allocate_common to get more details ).
+  // The opto fast path object allocation tracing occurs in the opto CallLeafNode,
+  // which has been broken by invalid top frame bci.
+  // To fix this, we get the top frame bci in opto compilation phase
+  // and pass it as parameter to runtime call. Our implementation will replace the invalid top
+  // frame bci with cached_top_frame_bci.
+  jint _cached_top_frame_bci;
+  jlong _alloc_count;
+  jlong _alloc_count_until_sample;
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply src/share/vm/jfr/recorder/access/jfrThreadData.hpp]: "_alloc_count" means the numbers of objects allocated by this thread so far, and then sampling will be triggered when _alloc_count == _alloc_count_until_sample

File: src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp (hunk 1 above)
Comment (D-D-H) [reply src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp]: the event parser

File: src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp (hunk 2 above)
Comment (D-D-H) [reply src/share/vm/jfr/recorder/checkpoint/constant/jfrTagSet.cpp]: done

File: src/share/vm/jfr/objectprofiler/objectProfiler.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
+#include "runtime/vmThread.hpp"
+#include "jfr/objectprofiler/objectProfiler.hpp"
+#include "jfr/utilities/jfrTryLock.hpp"
+#include "tracefiles/traceEventClasses.hpp"
+
+jint ObjectProfiler::_enabled = JNI_FALSE;
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply src/share/vm/jfr/objectprofiler/objectProfiler.cpp]: The value of this variable is updated by compiled code of C2, I don't recommend changing the type of this variable now, the current version is stable and verified on our production environment.

File: src/share/vm/jfr/recorder/access/jfrThreadData.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (luchsh) [reply src/share/vm/jfr/recorder/access/jfrThreadData.hpp]: _alloc_count --> _allocated_objs?

[the remark this answers is not in the document]
Comment (D-D-H) [reply src/share/vm/jfr/recorder/access/jfrThreadData.hpp]: I think "_alloc_count" is enough to express the meaning of it here.
And we can discuss more deeply before we contribute it to OpenJDK.

File: test/jfr/event/objectsprofiling/TestOptoObjectAllocationsSampling.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
+package jfr.event.objectsprofiling;
```
Comment (sanhong) [inline test/jfr/event/objectsprofiling/TestOptoObjectAllocationsSampling.java]: pls. add copyright header here.

Comment (D-D-H) [reply test/jfr/event/objectsprofiling/TestOptoObjectAllocationsSampling.java]: done

File: src/share/vm/jfr/objectprofiler/objectProfiler.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  }
+  OrderAccess::release_store((volatile jint*)&_enabled, JNI_TRUE);
+}
+
+void ObjectProfiler::stop(jlong event_id) {
+#ifndef PRODUCT
+  JfrTryLock try_lock(&_try_lock);
+  assert(try_lock.has_lock(), "Not allow contention");
+#endif
+  if (_enabled == JNI_FALSE) {
```
Comment (kuaiwei) [inline src/share/vm/jfr/objectprofiler/objectProfiler.cpp]: If _enabled is stored by release_store, it should be loaded with load_acquire. And why not declare _enabled as volatile?

File: src/share/vm/opto/macro.cpp [hunk 10]
```diff
@@ -1619,6 +1631,114 @@ void PhaseMacroExpand::expand_allocate_common(
...
+  IfNode* alloc_sample_enabled_if = (IfNode*)transform_later(new (C) IfNode(fast_oop_ctrl, alloc_sample_enabled_bool, PROB_MIN, COUNT_UNKNOWN));
+  Node* alloc_sample_enabled_ctrl = transform_later(new (C) IfTrueNode(alloc_sample_enabled_if));
+  Node* alloc_sample_enabled_mem = fast_oop_rawmem;
+  Node* alloc_sample_disabled_ctrl = transform_later(new (C) IfFalseNode(alloc_sample_enabled_if));
+  Node* alloc_sample_disabled_mem = fast_oop_rawmem;
+  Node* alloc_sample_enabled_region = transform_later(new (C) RegionNode(3));
+  Node* alloc_sample_enabled_region_phi_mem = transform_later(new (C) PhiNode(alloc_sample_enabled_region, Type::MEMORY, TypeRawPtr::BOTTOM));
+  enum { enabled_idx = 1,  disabled_idx = 2 };
+
+  // if_then
```
Comment (kuaiwei) [inline src/share/vm/opto/macro.cpp]: the 2 if_then comments are confusing. I think the 1st if_then can be "if _enabled then", the second if_then is "if sample_hit then"

File: src/share/vm/jfr/objectprofiler/objectProfiler.cpp (hunk 9 above)
Comment (D-D-H) [reply src/share/vm/jfr/objectprofiler/objectProfiler.cpp]: fixed

File: src/share/vm/opto/macro.cpp (hunk 10 above)
Comment (D-D-H) [reply src/share/vm/opto/macro.cpp]: fixed

**Pull request merged**